core/shared/src/main/scala/scalaz/zio/Queue.scala

```diff
@@ -3,156 +3,404 @@
+                      val toQueue  = capacity.fold(notTaken)(notTaken.takeRight)
+                      p.complete(false) -> Surplus(
+                        IQueue.empty.enqueue(toQueue.toList),
+                        IQueue.empty,
+                        hook
+                      )
+                    case Dropping =>
+                      p.complete(false) -> Surplus(
+                        IQueue.empty.enqueue(addToQueue.toList),
+                        IQueue.empty,
+                        hook
+                      )
+                  }
+
+              completeTakers *> complete -> surplus
+            } else
+              completeTakers *> p.complete(true) -> Deficit(deficitValues, hook)
         }
 
-      case (p, Surplus(values, putters)) =>
-        if (values.length < capacity && putters.isEmpty) {
-          (p.complete(()), Surplus(values.enqueue(a), putters))
-        } else {
-          (IO.now(false), Surplus(values, putters.enqueue((a, p))))
+      case (p, Surplus(values, putters, hook)) =>
+        val (addToQueue, surplusValues) =
+          capacity.fold((as, Iterable.empty[A]))(c => as.splitAt(c - values.size))
+        if (surplusValues.isEmpty)
+          (p.complete(true), Surplus(values.enqueue(addToQueue.toList), putters, hook))
+        else {
+          strategy match {
+            case BackPressure =>
+              (
+                IO.now(true),
+                Surplus(values.enqueue(addToQueue.toList), putters.enqueue(surplusValues -> p), hook)
+              )
+            case Sliding =>
+              (
+                p.complete(true),
+                capacity.fold(
+                  Surplus(values.enqueue(as.toList), putters, hook)
+                ) { c =>
+                  Surplus(
+                    values
+                      .takeRight(c - as.size)
+                      .enqueue(as.takeRight(c).toList),
+                    putters,
+                    hook
+                  )
+                }
+              )
+            case Dropping =>
+              (
+                p.complete(false),
+                Surplus(values.enqueue(addToQueue.toList), putters, hook)
+              )
+          }
         }
+
+      case (p, state @ Shutdown) => (p.interrupt, state)
     }
 
-    val release: (Boolean, Promise[Nothing, Unit]) => IO[Nothing, Unit] = {
-      case (_, p) => removePutter(p)
+    val release: (Boolean, Promise[Nothing, Boolean]) => IO[Nothing, Unit] = {
+      case (_, p) => p.poll.void <> removePutter(p)
     }
 
-    Promise.bracket[Nothing, State[A], Unit, Boolean](ref)(acquire)(release)
+    Promise.bracket[Nothing, State[A], Boolean, Boolean](ref)(acquire)(release)
   }
 
+  /**
+   * Adds a shutdown hook that suspends until `shutdown` is called.
+   * The hook will resume immediately when the queue is already shutdown.
+   */
+  final def awaitShutdown(io: IO[Nothing, Unit]): IO[Nothing, Unit] =
+    Promise
```

Comment:
There is no need for the `io` parameter. Please see the implementation of `awaitShutdown` that I provided previously.

In this model, the returned `IO` will not resume until shutdown is complete. So there is no need for `io`. If I didn't make any mistakes (probably I made some 😄), you should be able to copy/paste the definition of `awaitShutdown` that I provided.

Reply:
@jdegoes I copy/paste the definition of `awaitShutdown` (it's the same) but the `io` made me confused :/ so I added it in parameter.
how can we add a hook if we don't need an `io` in parameter ?

Reply:
You can see in my definition, I did:

```scala
val hook = promise.complete(())
```

The hook completes the promise whose `get` is returned from by `awaitShutdown`.

Reply:
@wi101 Oops I made an error, it should be:

```scala
final def awaitShutdown: IO[Nothing, Unit] = 
  Promise.make[Nothing, Unit].flatMap(promise => {
    val io = promise.complete(())
    IO.flatten(ref.modify {
      case Deficit(takers, hook)         => IO.unit -> Deficit(takers, hook *> io)
      case Surplus(queue, putters, hook) => IO.unit -> Surplus(queue, putters, hook *> io)
      case state @ Shutdown              => io      -> state
    }) *> promise.get
  })
```

Reply:
I saw that and in your definition there is `io`.
OK, so the implementation would be like this:
```scala
final def awaitShutdown: IO[Nothing, Unit] = 
  Promise.make[Nothing, Unit].flatMap(promise => {
    val hook = promise.complete(())
    IO.flatten(ref.modifySome(IO.unit) {
     case state @ Shutdown              => hook -> state
    }) *> promise.get
  })
```
Is that correct?

Reply:
Ah ok now sounds clear, @jdegoes thanks

Reply:
Sorry for the confusion. 😊